icicle-compiler/src/Icicle/Runtime/Data/Array.hs

```diff
@@ -297,9 +297,24 @@ grow pool arrayOld@(Array ptrOld) lengthNew = do
     !capacityNew =
       calculateCapacity lengthNew
 
-  if capacityOld >= capacityNew then do
+  if lengthOld > lengthNew then do
+    let
+      deadSize =
+        (lengthOld - lengthNew) * elementSize
+
+      deadPtr =
+        unsafeElementPtr arrayOld `plusPtr`
+        fromIntegral (lengthNew * elementSize)
+
+    c_memset deadPtr 0 (fromIntegral deadSize)
```

Comment:
I actually found a bug, it was expected that when you grow an array, all the newly added elements start off as `0`, this wasn't the case if you shrink an array then grow it again:

```hs
Var 2 = NewArray (ArrayLength 1)
Var 4 = WriteArray (ArrayIndex 0) (-1) (Var 2)
Var 5 = GrowArray (ArrayLength 0) (Var 2)
Var 6 = GrowArray (ArrayLength 1) (Var 5)
Var 7 = ReadArray (ArrayIndex 0) (Var 6) -- reads the -1 value!
```